### PR TITLE
VoiceTypeAssetLookup optimisations

### PR DESCRIPTION
**Pull request by kieranknowles1**
Optimizations and refactoring of VoiceTypeAssetLookup

- Reduce usage of HashSet during prepare
- Build lookups from winning overrides
- Cache voice type -> speakers for cases where the whole voice type is used
- Use Lazy for child NPCs, scene aliases, and shared info usages
- Use enumerable for SpeakerVoices value as most NPCs only have one voice type.

Scene aliases and shared info usages may benefit further from a usage cache, this is not implemented here as it would be a breaking change.

Overall, this gives a 1.8x speedup and reduction in allocations when preparing the lookup against the vanilla masters and a
45x speedup with 5x reduction in allocations when processing GetIsVoiceType conditions.

---
Before:
| Method                      | Mean         | Error      | StdDev     | Gen0    | Gen1 | Gen2  | Allocated |
|---------------------------- | ------------:|-----------:|-----------:|--------:|-------:|-:|----------:|
| PrepareWithDefaultMasters | 261.9 ms | 5.21 ms |  4.87 ms | 17000.0000 | 5000.0000 | 1000.0000 | 282.92 MB |
| LookupVoicesSpecificSpeaker |     2.384 us |  0.0473 us |  0.0790 us |  0.4158 |      - | - |    6.8 KB |
| LookupVoicesGenericSpeaker  | 5,460.695 us | 49.8194 us | 41.6015 us | 15.6250 | 7.8125 | - | 379.58 KB |

After:
| Method                      | Mean       | Error     | StdDev    | Gen0   | Gen1 | Gen2  | Allocated |
|------------------------ |-----------:|----------:|----------:|-------:|-------:|-:|----------:|
| PrepareWithDefaultMasters | 144.8 ms | 2.53 ms | 3.19 ms | 9000.0000 | 2000.0000 |        - | 157.56 MB |
| LookupVoicesSpecificSpeaker |   2.178 us | 0.0197 us | 0.0175 us | 0.4005 |      - | - |   6.59 KB |
| LookupVoicesGenericSpeaker | 119.337 us | 0.6232 us | 0.5204 us | 4.0283 | 0.7324 | - |  67.18 KB |

### CODE REVIEW

File: Mutagen.Bethesda.Skyrim/Records/Assets/VoiceType/VoiceTypeAssetLookup.cs [hunk 1]
```diff
@@ -12,14 +12,18 @@ public class VoiceTypeAssetLookup : IAssetCacheComponent
 
     //Databases
     private readonly Dictionary<ModKey, HashSet<string>> _defaultVoiceTypes = new();
-    private readonly Dictionary<FormKey, HashSet<string>> _speakerVoices = new();
+    // NPC -> Voice types
+    private readonly Dictionary<FormKey, IEnumerable<string>> _speakerVoices = new();
```
Comment (Noggog) [inline Mutagen.Bethesda.Skyrim/Records/Assets/VoiceType/VoiceTypeAssetLookup.cs:16]: Why make the value type here IEnumerable<string> when we're filling it with a HashSet<string> on line 67?

Doesn't matter too much, since it's all private.. but seems like an unnecessary change

File: Mutagen.Bethesda.Tests/Benchmarks/VoiceTypeAssetLookupBenchmark.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        var env = GameEnvironment.Typical.Builder<ISkyrimMod, ISkyrimModGetter>(GameRelease.SkyrimSE)
+            .Build();
+        LinkCache = env.LinkCache;
+        Lookup = new();
+        var assetCache = LinkCache.CreateImmutableAssetLinkCache();
+        Lookup.Prep(assetCache);
+
+        ResponsesSpecific = LinkCache.Resolve<IDialogResponsesGetter>(FormKey.Factory("0D5533:Skyrim.esm"));
+        ResponsesGeneric = LinkCache.Resolve<IDialogResponsesGetter>(FormKey.Factory("0DBA2A:Skyrim.esm"));
```
Comment (Noggog) [inline Mutagen.Bethesda.Tests/Benchmarks/VoiceTypeAssetLookupBenchmark.cs:54]: Just want to confirm these benchmark tests are just using skyrim base master content and don't rely on other mods you might have installed to measure properly

File: Mutagen.Bethesda.Skyrim/Records/Assets/VoiceType/VoiceTypeAssetLookup.cs [hunk 3]
```diff
@@ -807,156 +815,123 @@ private IEnumerable<string> GetVoiceTypes(FormKey speaker)
...
-        if ((npc.Configuration.TemplateFlags & NpcConfiguration.TemplateFlag.Traits) == 0 && !npc.Race.IsNull)
-        {
-            return new HashSet<FormKey> { npc.Race.FormKey };
-        }
+        // TODO: Account for templates
+        if (!npc.Template.IsNull && npc.Configuration.TemplateFlags.HasFlag(NpcConfiguration.TemplateFlag.Traits))
+            return [];
 
-        return [];
+        return [npc.Race];
```
Comment (Noggog) [inline Mutagen.Bethesda.Skyrim/Records/Assets/VoiceType/VoiceTypeAssetLookup.cs:934]: Looked like the lhs side checked if the npc.Race was null first?  Is that removal intentional?  If so, maybe we update some of the unit tests to reflect

Might be good to brush up on those in general.  I remember at the time of making this (or maybe some similar asset logic) didn't have as many tests as desired.   Could be good to lock in the behavior with the tests first, then apply the optimization logic

File: Mutagen.Bethesda.Skyrim/Records/Assets/VoiceType/VoiceTypeAssetLookup.cs [hunk 4]
```diff
@@ -807,156 +815,123 @@ private IEnumerable<string> GetVoiceTypes(FormKey speaker)
...
     {
-        if ((npc.Configuration.TemplateFlags & NpcConfiguration.TemplateFlag.Stats) == 0 && !npc.Class.IsNull)
-        {
-            return new HashSet<FormKey> { npc.Class.FormKey };
-        }
+        if (!npc.Template.IsNull && npc.Configuration.TemplateFlags.HasFlag(NpcConfiguration.TemplateFlag.Stats))
+            return GetClasses(npc.Template);
 
-        return npc.Template.IsNull ? new HashSet<FormKey>() : GetClasses(npc.Template);
+        return [npc.Class];
```
Comment (Noggog) [inline Mutagen.Bethesda.Skyrim/Records/Assets/VoiceType/VoiceTypeAssetLookup.cs:895]: As mentioned in the other comment, this seems to have logic differences.  

It feels like the old logic returned an empty set if the class and template was null.  Whereas this will return npc.Class even if it's null?

I wouldn't be too worried, except no unit tests needed to be changed, which suggests the behavior wasn't really flexed by them too well in the first place

What IS the intention?  Is it correct that we return empty if the class is null or not?   I'm not familiar enough with the goals to know the answer... but solidifying the choice in unit tests will at least solidify the intention one way or the other

File: Mutagen.Bethesda.Skyrim/Records/Assets/VoiceType/VoiceTypeAssetLookup.cs [hunk 5]
```diff
@@ -807,156 +815,123 @@ private IEnumerable<string> GetVoiceTypes(FormKey speaker)
     }
 
     #region Voice Parser
-    private HashSet<string> GetVoiceTypes(INpcGetter npc)
+    private IEnumerable<string> GetVoiceTypes(INpcGetter npc)
```
Comment (Noggog) [inline Mutagen.Bethesda.Skyrim/Records/Assets/VoiceType/VoiceTypeAssetLookup.cs:818]: I dont mind the HashSet return value here.  It's private.  And we are returning hash sets. And they are more "useful" than straight IEnumerable.

I think dropping the type back to IEnumerable is only interesting in public interfaces where we might change our minds later on what the backing collection is.  N/A for private functions

**Review by Noggog — APPROVED** [review]
Thanks for the PR!   Great to see things getting optimized.

My main gripe is mainly a past sin where we didn't finish off writing unit tests to go with the logic.   This makes followups like this harder as we have to evaluate if the improvements changed any behaviors.    Are there edge case differences introduced in the change?   Is it faster just because it's not actually doing the same job?  Hard to tell/prove.

I do see -some- tests in the VoiceTypeAssetLookupTestSkyrim.cs file.  But I also noticed some logic changes that suggest the tests aren't quite complete.   Maybe/hopefully some of the differences i spotted are non-issues.  But I think it's worth considering upgrading the tests as well while we're here.   Even defining the tests help communicate intent very well.  Do we INTEND for X to be returned when Y is true, etc.

**Comment by kieranknowles1** [comment]
> Thanks for the PR! Great to see things getting optimized.
> 
> My main gripe is mainly a past sin where we didn't finish off writing unit tests to go with the logic. This makes followups like this harder as we have to evaluate if the improvements changed any behaviors. Are there edge case differences introduced in the change? Is it faster just because it's not actually doing the same job? Hard to tell/prove.
> 
> I do see -some- tests in the VoiceTypeAssetLookupTestSkyrim.cs file. But I also noticed some logic changes that suggest the tests aren't quite complete. Maybe/hopefully some of the differences i spotted are non-issues. But I think it's worth considering upgrading the tests as well while we're here. Even defining the tests help communicate intent very well. Do we INTEND for X to be returned when Y is true, etc.

The main intentional change is to only consider winning overrides. I.e., if a plugin changes an NPCs voice, then only include the new voice.

You're right that unit tests should be extended, it's probably smart to do this in a separate branch based on the original implementation.

File: Mutagen.Bethesda.Skyrim/Records/Assets/VoiceType/VoiceTypeAssetLookup.cs (hunk 5 above)
Comment (kieranknowles1) [reply Mutagen.Bethesda.Skyrim/Records/Assets/VoiceType/VoiceTypeAssetLookup.cs:818]: Most non-generic NPCs have exactly one voice type, so using a HashSet seems overkill.

With the addition of _voiceSpeakers for GetIsVoiceType conditions _speakerVoices may be redundant. I'll have to do some more testing to see if an enumerable is better and if removing _speakerVoices is worth the memory savings. The other _fieldNPCs lookups would probably benefit from being lazy, but less so from a usage cache since they can be inherited.

File: Mutagen.Bethesda.Tests/Benchmarks/VoiceTypeAssetLookupBenchmark.cs (hunk 2 above)
Comment (kieranknowles1) [reply Mutagen.Bethesda.Tests/Benchmarks/VoiceTypeAssetLookupBenchmark.cs:54]: I'm using mod organiser, so my testing environment consists only of the base masters + free CC.

File: Mutagen.Bethesda.Skyrim/Records/Assets/VoiceType/VoiceTypeAssetLookup.cs (hunk 5 above)
Comment (Noggog) [reply Mutagen.Bethesda.Skyrim/Records/Assets/VoiceType/VoiceTypeAssetLookup.cs:818]: Ah gotcha.  Yeah, i see now where the logic can sometimes return not a hash set

**Comment by kieranknowles1** [comment]
Closing this as #681 covers much of the same areas and may be impractical to merge.
